rotkehlchen/chain/evm/decoding/metamask/decoder.py

```diff
@@ -0,0 +1,177 @@
+    from rotkehlchen.user_messages import MessagesAggregator
+
+
+class MetamaskCommonDecoder(DecoderInterface):
+
+    def __init__(
+            self,
+            evm_inquirer: 'EvmNodeInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+            router_address: str,
+            gnosis_safe_proxy: bytes,
+            native_currency: Asset,
+    ) -> None:
+        super().__init__(evm_inquirer, base_tools, msg_aggregator)
+        self.router_address = string_to_evm_address(router_address)
+        self.gnosis_safe_proxy = gnosis_safe_proxy
+        self.native_currency = native_currency
+
+    def _decode_swap(self, context: DecoderContext) -> DecodingOutput:
+        """
+        This function is used to decode the swap done by metamask.
+        By reading the contract code the spending address and the receiving address
+        seem to be the same, so it doesn't check for that.
+        """
+
+        if context.tx_log.topics[0] != SWAP_SIGNATURE:
+            return DEFAULT_DECODING_OUTPUT
+
+        sender = hex_or_bytes_to_address(context.tx_log.topics[2])
+        if not self.base.is_tracked(sender):
+            return DEFAULT_DECODING_OUTPUT
+
+        out_event = in_event = None
+        for event in context.decoded_events:
+            asset_symbol = event.asset.resolve_to_asset_with_symbol().symbol
+            if (
+                event.event_type == HistoryEventType.TRADE and
+                event.event_subtype == HistoryEventSubType.SPEND or
+                event.event_type == HistoryEventType.SPEND and
+                event.event_subtype == HistoryEventSubType.NONE and
+                event.location_label == sender
+            ):
+                # find the send event
+                event.counterparty = CPT_METAMASK
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.SPEND
+                event.notes = f'Swap {event.balance.amount} {asset_symbol} in {CPT_METAMASK} from {event.location_label}'  # noqa: E501
+                event.address = self.router_address
+                out_event = event
+            elif event.event_type == HistoryEventType.RECEIVE and event.location_label == sender:
+                # find the receive event
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.RECEIVE
+                event.counterparty = CPT_METAMASK
+                event.notes = f'Receive {event.balance.amount} {asset_symbol} from {CPT_METAMASK} swap in {event.location_label}'  # noqa: E501
+                event.address = self.router_address
+                # use this index as the event may be a native currency transfer
+                # and appear at the start
+                event.sequence_index = context.tx_log.log_index
+                in_event = event
+                to_amount = event.balance.amount
```

Comment:
This is only defined here. If the logic, for whatever reason, does not get in here and then proceeds to line 143, then `to_amount` will be undefined. You need to either set it to `ZERO` before the loop or set to None and after the loop check if it's still None and error out.